triton-vm/src/bfield_codec.rs

```diff
@@ -0,0 +1,663 @@
+    fn decode(str: &[BFieldElement]) -> Result<Box<Self>, Box<dyn Error>> {
+        if str.len() != 1 {
+            return Err(BFieldCodecError::boxed(
+                "trying to decode more or less than one BFieldElements as one BFieldElement",
+            ));
+        }
+        let element_zero = str[0];
+        Ok(Box::new(element_zero))
+    }
+
+    fn encode(&self) -> Vec<BFieldElement> {
+        [*self].to_vec()
+    }
+}
+
+impl BFieldCodec for XFieldElement {
+    fn decode(str: &[BFieldElement]) -> Result<Box<Self>, Box<dyn Error>> {
+        if str.len() != 3 {
+            Err(BFieldCodecError::boxed(
+                "trying to decode slice of not 3 BFieldElements into XFieldElement",
+            ))
+        } else {
+            Ok(Box::new(XFieldElement {
+                coefficients: str.try_into().unwrap(),
+            }))
+        }
+    }
+
+    fn encode(&self) -> Vec<BFieldElement> {
+        self.coefficients.to_vec()
+    }
+}
+
+impl BFieldCodec for Digest {
+    fn decode(str: &[BFieldElement]) -> Result<Box<Self>, Box<dyn Error>> {
+        if str.len() != DIGEST_LENGTH {
+            Err(BFieldCodecError::boxed(
+                "trying to decode slice of not DIGEST_LENGTH BFieldElements into Digest",
+            ))
+        } else {
+            Ok(Box::new(Digest::new(str.try_into().unwrap())))
+        }
+    }
+
+    fn encode(&self) -> Vec<BFieldElement> {
+        self.to_sequence()
+    }
+}
+
+// impl<T: BFieldCodec> BFieldCodec for Vec<T> {
+//     fn decode(str: &[BFieldElement]) -> Result<Box<Self>, Box<dyn Error>> {
+//         let mut vect: Vec<T> = vec![];
+//         let mut index = 0;
+//         while index < str.len() {
+//             // we don't know ahead of time how wide each T-element is
+//             // so every element in the list is length-prepended
+//             let len = str[index].value();
+//             if str.len() < index + 1 + len as usize {
+//                 return Err(BFieldCodecError::boxed(
+//                     "cannot decode vec of Ts because of erroneous length prepending",
+//                 ));
+//             }
+//             let substr = &str[(index + 1)..(index + 1 + len as usize)];
+//             let decoded = T::decode(substr);
+//             if let Ok(t) = decoded {
+//                 vect.push(*t);
+//             } else {
+//                 return Err(BFieldCodecError::boxed("cannot decode T element in vec"));
+//             }
+
+//             index += 1 + len as usize;
+//         }
+//         Ok(Box::new(vect))
+//     }
+
+//     fn encode(&self) -> Vec<BFieldElement> {
+//         let mut str = vec![];
+//         for elem in self.iter() {
+//             let mut substr = elem.encode();
+//             str.push(BFieldElement::new(substr.len().try_into().expect("Generic parameter encoding length (as BFieldElements) does not fit into BFieldElement")));
+//             str.append(&mut substr);
+//         }
+//         str
+//     }
+// }
```

Comment:
Historical relic, fixed in the next commit.

We can't use this generic trait implementation because the serialization length of `T`-elements is not known. Therefore you would need to prepend the length. So a vector of `BFieldElements` `[3, 4, 5]` would be encoded as `1 3 1 4 1 5`. By switching to specific trait implementations we can avoid this factor 2 overhead.